Refactor DynamicGraph to use more efficient algorithms and data structures.

steveprentice opened the pull request:
DynamicGraph now organizes its edges in two maps of maps so it can quickly
lookup edges to or from any node to any other node. Previously, finding
edges, especially when searching by the "to" node, would require iterating
over all nodes. This caused an exponential performance curve resulting in
the LotsOfEdgesTest not being able to finish in any practical length of time.

Fixes #1710 

In #1710, we talked about taking a different approach to solving this problem, but in the end, I decided not to take that approach and keep DynamicGraph working pretty much the same way as before. The changes to DynamicGraph are mostly geared toward using a different data structure to manage the edges and adapting the code to take advantage of those data structures. The result is a massive performance boost without any additional memory usage (because the old code also kept two copies of the edges – m_edges and m_allEdges).

### Did you remember to?

- [X] Add test case(s)
- [X] Update `CHANGES.txt`

steveprentice commented on src/test/java/org/testng/internal/DynamicGraphHelperTest.java:
```diff
@@ -40,16 +41,14 @@
-        for (List<DynamicGraph.Edge<ITestNGMethod>> edge : graph.getEdges().values()) {
-            assertThat(edge).isEmpty();
-        }
+        assertThat(graph.getEdges().isEmpty());
```
All changes to these tests were to adapt the tests to the new data structure in DynamicGraph. No functional changes to the tests were made.

juherr replied:
`assertThat(graph.getEdges()).isEmpty()` looks better

steveprentice commented on src/test/java/org/testng/internal/DynamicGraphTest.java:
```diff
@@ -289,4 +290,59 @@ public void testDuplicationFunctionality() {
+    // Should only be one edge with weight of 2 now.
+    assertThat(dg.getEdges().values().size()).isEqualTo(1);
+    assertThat(dg.getEdges().get("a").get("b")).isEqualTo(2);
+  }
```
The above three tests were not covered previously but are important to how DynamicGraph works.

juherr replied:
:+1:

juherr commented:
A test is failing on Java7 (`test.ignore.IgnoreTest » ignore_class_should_not_run_tests`), check https://scans.gradle.com/s/3kejrpmt5q33u/tests/szvlqherh4whi-ymfnarya5ipg4 or https://scans.gradle.com/s/imry4ts5xuyqy/tests/szvlqherh4whi-ymfnarya5ipg4

I run Travis twice and the issue still persists.

steveprentice commented:
@juherr, that's an issue on master - check the latest master builds in Travis. I rebased my branch on the latest master, so that's why this PR is picking up that issue. 

E.g. https://travis-ci.org/cbeust/testng/jobs/351711735, has the same failure specific to jdk7. 
```
test.ignore.IgnoreTest.ignore_class_should_not_run_tests()
```

juherr commented:
@steveprentice Ok, thanks :)

juherr commented on src/main/java/org/testng/internal/DynamicGraph.java:
```diff
@@ -1,57 +1,36 @@
-    }
-  }
-
   public DynamicGraph() {
```
The constructor may be removed now

juherr commented on src/main/java/org/testng/internal/DynamicGraph.java:
```diff
@@ -69,75 +48,105 @@ public boolean addNode(T node) {
+  }
+
+
+  private void addEdgeToMap(Map<T, Map<T, Integer>> map, T n1, T n2, int weight) {
```
util method, could be static

juherr commented on src/main/java/org/testng/internal/DynamicGraph.java:
```diff
@@ -69,75 +48,105 @@ public boolean addNode(T node) {
+   * @param node Node to remove.
+   */
+  private void removeNode(T node) {
+    Map<T, Integer> incomingEdges = m_incomingEdges.get(node);
```
Could be:

```java
Map<T, Integer> incomingEdges = m_incomingEdges.remove(node);
Map<T, Integer> outgoingEdges = m_outgoingEdges.remove(node);
```

juherr commented on src/main/java/org/testng/internal/DynamicGraph.java:
```diff
@@ -69,75 +48,105 @@ public boolean addNode(T node) {
+    if (incomingEdges != null) {
+      for (T from : incomingEdges.keySet()) {
+        Map<T, Integer> edges = m_outgoingEdges.get(from);
+        edges.remove(node);
```
edges may be null

juherr commented on src/main/java/org/testng/internal/DynamicGraph.java:
```diff
@@ -69,75 +48,105 @@ public boolean addNode(T node) {
-    for (Edge<T> e : edges.get(edge.to)) {
-      if (e.to.equals(edge.from)) {
-        return e;
+    if (outgoingEdges != null) {
```
Could you add a one line comment on the 2 blocks which explains what you are doing?

juherr commented on src/main/java/org/testng/internal/DynamicGraph.java:
```diff
@@ -69,75 +48,105 @@ public boolean addNode(T node) {
-    // if all nodes have dependencies, then we can ignore the lowest one
-    if (result.isEmpty()) {
-      int lowestPriority = getLowestEdgePriority(m_nodesReady);
+    result.addAll(difference(m_nodesReady, m_outgoingEdges.keySet()));
```
Maybe:

```java
List<T> result = Lists.newArrayList(difference(m_nodesReady, m_outgoingEdges.keySet()));
```

juherr commented:
Fyi, the last time I fix the graph, I broke the parallel feature.
In fact, `getFreeNodes()` is supposed to return all methods which should be run at the same time. But currently, only one method is returned because all methods have edges. 
Maybe we should remove some of them in some specific case but I didn't find a good approach.

BTW, as you have a good knowledge of this part now, maybe you'll have an idea :)

steveprentice commented:
@juherr, thanks for the review. I'll get it all fixed it shortly. Regarding the parallel case, can you provide a test case? I'm not completely sure I understand. My understanding was that during parallel execution the preserve order edges don't get added at all.

juherr commented:
> Regarding the parallel case, can you provide a test case?

It will be difficult. I worked on it a long time ago. But I remember last time I checked that `getFreeNodes` never return more than 1 method.

steveprentice commented:
@juherr, I've updated the PR to address all your review comments. Also:

* Moved the edge management stuff into an DynamicGraph.Edges internal class to help isolate that logic. Before, the edge management stuff was spread throughout DynamicGraph. I think this makes it easier to comprehend the data structures behind the new approach.
* Added an explicit test that covers the parallel case you talk about. getFreeNodes() is working as expected in that case.

steveprentice commented:
Huh. But that broke a bunch of tests. Looking into that.

steveprentice commented:
Ok. Back to passing on JDK8 and JDK9 builds. JDK7 builds are still failing because of master.

steveprentice commented:
Added one last commit (hopefully) to make DynamicGraphTest assert the order of the nodes returned by getFreeNodes(). This would have saved me from the previous bad commit that ended up changing that order as I was only running the DynamicGraphTest tests and not the whole suite.

steveprentice commented:
Any concerns about merging this?
